staging/src/k8s.io/cloud-provider/controllers/service/controller_test.go

```diff
@@ -615,37 +620,87 @@ func TestNodeChangesInExternalLoadBalancer(t *testing.T) {
 			},
 			worker:                3,
 			nodeListerErr:         nil,
-			expectedRetryServices: []*v1.Service{},
+			expectedRetryServices: []string{},
 		},
 		{
 			desc:                  "error occur during sync",
 			nodes:                 []*v1.Node{node1, node2, node3, node4},
 			expectedUpdateCalls:   []fakecloud.UpdateBalancerCall{},
 			worker:                3,
 			nodeListerErr:         fmt.Errorf("random error"),
-			expectedRetryServices: services,
+			expectedRetryServices: serviceNames,
 		},
 		{
 			desc:                  "error occur during sync with 1 workers",
 			nodes:                 []*v1.Node{node1, node2, node3, node4},
 			expectedUpdateCalls:   []fakecloud.UpdateBalancerCall{},
 			worker:                1,
 			nodeListerErr:         fmt.Errorf("random error"),
-			expectedRetryServices: services,
+			expectedRetryServices: serviceNames,
 		},
 	} {
 		t.Run(tc.desc, func(t *testing.T) {
 			ctx, cancel := context.WithCancel(context.Background())
 			defer cancel()
 			controller.nodeLister = newFakeNodeLister(tc.nodeListerErr, tc.nodes...)
-			servicesToRetry := controller.updateLoadBalancerHosts(ctx, services, tc.worker)
-			compareServiceList(t, tc.expectedRetryServices, servicesToRetry)
+			var servicesToRetry []string
+			for key := range controller.updateLoadBalancerHosts(ctx, services, tc.worker) {
+				servicesToRetry = append(servicesToRetry, key)
+			}
+			compareServiceNameList(t, tc.expectedRetryServices, servicesToRetry)
 			compareUpdateCalls(t, tc.expectedUpdateCalls, cloud.UpdateCalls)
 			cloud.UpdateCalls = []fakecloud.UpdateBalancerCall{}
 		})
 	}
 }
 
+// TestServiceUpdateWithLatestSpec tests if Services are updated with stale specs or the latest ones.
+// There's one test here. service0 has ServicePort: 80 at the beginning. Then a random error is
+// passed to the nodeLister, so the first updateServices() will fail. Next, service0's ServicePort is
+// changed to 8080 and we retry updateServices(). Finally, the test checks if the cached Service has
+// latest ServicePort: 8080.
+func TestServiceUpdateWithLatestSpec(t *testing.T) {
+	controller, _, _ := newController()
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	node0 := &v1.Node{ObjectMeta: metav1.ObjectMeta{Name: "node0"}, Status: v1.NodeStatus{Conditions: []v1.NodeCondition{{Type: v1.NodeReady, Status: v1.ConditionTrue}}}}
+	nodeListerErr := fmt.Errorf("random error to make services retry")
+	controller.nodeLister = newFakeNodeLister(nodeListerErr, node0)
+
+	service0 := newService("s0", "000", v1.ServiceTypeLoadBalancer)
+	service0.Spec.Ports = []v1.ServicePort{{Port: 80}}
+	serviceKey0 := fmt.Sprintf("%s/%s", service0.GetObjectMeta().GetNamespace(), service0.GetObjectMeta().GetName())
+	cachedSvc := controller.cache.getOrCreate(serviceKey0)
+	cachedSvc.state = service0
+	controller.cache.set(serviceKey0, cachedSvc)
+
+	updateServices := func() []string {
```

Comment:
The block of code inside this function is the code is similar to what we are trying to test inside nodeSyncInternal. I think it would make more sense to refactor this test to actually call nodeSyncInternal()

Reply:
I removed the test for the following reason. Whether or not to call `nodeSyncInternal()` in the test cannot tell you if the updated spec is used. All you are dealing with is the `controller.servicesToUpdate`, which dones't contain Service spec info. The original test code just checks value in cache which is always correct.

Reply:
@andrewsykim hello, could you share your idea?

Reply:
@andrewsykim I created a PR to keep the UT but changed the `nodeSyncInternal()`. Which do you prefer?
https://github.com/kubernetes/kubernetes/pull/109062